### PR TITLE
PMax Assets - Add portrait marketing images

### PR DESCRIPTION
**Pull request by jorgemd24**
### Changes proposed in this Pull Request:

Closes part of https://github.com/woocommerce/google-listings-and-ads/issues/1780

While testing with the Ads API, we found that the maximum number of marketing images, it is the combination of square + land space + portrait, so the maximum of those three types of images can not exceed 20 images. After this finding, we decided to add the Portrait Marketing images to the MVP, until now we were working only with landscape and square types. For more context see here: pcTzPl-12o-p2#comment-2228

This PR adds the following functionalities:
 - Fetch Portrait Marketing images from an asset group.
 - Update Portrait Marketing for a specific asset group.
 - Adds Portrait Marketing image in the asset suggestions page.
 - Limits the number of marketing image suggestions to 20 (10 landscape + 5 square + 5 portrait). These limits can be changed in the future and are not confirmed.

#### Update & Fetch a landscape image:

1. As the branch feature/pmax-assets is using some new imports from the GoogleAds Library, you may need to remove your vendor folder `rm -Rf vendor/*` and install again the dependencies using `composer install`.
1. Choose an Ads Campaign and get the campaign id.
2. Make a request to `GET gla/ads/campaigns/asset-groups?campaign_id=YOUR_CAMPAIGN_ID`, and get the asset group id returned in the response (the endpoint returns only one asset group).
3. Make the following request`PUT gla/ads/campaigns/asset-groups/YOUR_ASSET_GROUP_ID` with the below body. The operation should be successful:
<details>
<summary> Click here if you have an empty asset group (without headlines, descriptions,etc)</summary>

```
{
   "assets": [
       {
           "id" : null,
           "field_type": "headline",
           "content": "headline 1"
       },
       {
           "id" : null,
           "field_type": "headline",
           "content": "headline 2"
       },
       {
           "id" : null,
           "field_type": "headline",
           "content": "headline 3"
       },
       {
           "id" : null,
           "field_type": "long_headline",
           "content": "long headline 1"
       },
       {
           "id" : null,
           "field_type": "description",
           "content": "description 1"
       },
       {
           "id" : null,
           "field_type": "description",
           "content": "description 2"
       },
       {
           "id" : null,
           "field_type": "business_name",
           "content": "business name test"
       },       
       {
           "id" : null,
           "field_type": "square_marketing_image",
           "content": "https://wee-hedgehog.jurassic.ninja/wp-content/uploads/2023/01/test_square.png"
       },  
       {
           "id" : null,
           "field_type": "marketing_image",
           "content": "https://wee-hedgehog.jurassic.ninja/wp-content/uploads/2023/01/test_landscape.jpeg"
       },
       {
        "id" : null,
        "field_type": "portrait_marketing_image",
        "content": "https://wee-hedgehog.jurassic.ninja/wp-content/uploads/2023/01/test_portrait.jpeg"
        },       
       {
           "id" : null,
           "field_type": "logo",
           "content": "https://wee-hedgehog.jurassic.ninja/wp-content/uploads/2023/01/test_logo.jpeg"
       }
   ]
}
```
</details>

<details>
<summary> Click here if you have an asset group that already contains the mandatory fields (headlines, descriptions,etc) </summary>

```
{
   "assets": [
       {
        "id" : null,
        "field_type": "portrait_marketing_image",
        "content": "https://wee-hedgehog.jurassic.ninja/wp-content/uploads/2023/01/test_portrait.jpeg"
        } 
   ]
}
```

</details>

4. Make a request to `GET gla/ads/campaigns/asset-groups?campaign_id=YOUR_CAMPAIGN_ID`, you should see the image updated inside `portrait_marketing_image` 

#### Portrait images suggestions:

1. Choose a post ID or term ID that contains some images.
2. Make a request to `wc/gla/assets/suggestions?id=YOUR_ID&type=POST|TERM`
3. If the image is bigger than the minimum requirement for portrait images (480 x 600px), you should see an entry `portrait_marketing_image` containing images with an aspect ratio of 4:5.

### Additional details:
 - [Pmax Assets requirements.](https://developers.google.com/google-ads/api/docs/performance-max/assets)
 - To read more about the endpoint to manipulate the assets (create/update/delete)  see this PR: https://github.com/woocommerce/google-listings-and-ads/pull/1832
 - The number of suggestions images for each type (landscape, square and portrait) is not confirmed yet and could be changed in the future.

 

### Changelog entry

### CODE REVIEW

**Comment by puntope** [comment]
One thing I noticed (beyond this PR) is that when doing an edit with empty body 

`PUT gla/ads/campaigns/asset-groups/YOUR_ASSET_GROUP_ID`

Still get a success message. I was expecting to get some kind of error.

File: src/Ads/AssetSuggestionsService.php [hunk 1]
```diff
@@ -286,6 +299,7 @@ protected function get_post_assets( int $id ): array {
 			AssetFieldType::BUSINESS_NAME            => get_bloginfo( 'name' ),
 			AssetFieldType::SQUARE_MARKETING_IMAGE   => $marketing_images[ self::SQUARE_MARKETING_IMAGE_KEY ] ?? [],
 			AssetFieldType::MARKETING_IMAGE          => $marketing_images [ self::MARKETING_IMAGE_KEY ] ?? [],
+			AssetFieldType::PORTRAIT_MARKETING_IMAGE => $marketing_images [ self::PORTRAIT_MARKETING_IMAGE_KEY ] ?? [],
```
Comment (puntope) [inline src/Ads/AssetSuggestionsService.php]: 💅 What do you thing of refactor this a bit?

Maybe get a common fields
```

             AssetFieldType::LOGO                     => $this->get_logo_images(),
			AssetFieldType::BUSINESS_NAME            => get_bloginfo( 'name' ),
			AssetFieldType::SQUARE_MARKETING_IMAGE   => $marketing_images[ self::SQUARE_MARKETING_IMAGE_KEY ] ?? [],
			AssetFieldType::MARKETING_IMAGE          => $marketing_images [ self::MARKETING_IMAGE_KEY ] ?? [],
			AssetFieldType::PORTRAIT_MARKETING_IMAGE => $marketing_images [ self::PORTRAIT_MARKETING_IMAGE_KEY ] ?? [],
		    AssetFieldType::CALL_TO_ACTION_SELECTION => null,	
```

And add after the term/post specific ones

File: src/Ads/AssetSuggestionsService.php [hunk 2]
```diff
@@ -333,6 +347,7 @@ protected function get_term_assets( int $id ): array {
...
 			AssetFieldType::CALL_TO_ACTION_SELECTION => null,
```
Comment (puntope) [inline src/Ads/AssetSuggestionsService.php]: See https://github.com/woocommerce/google-listings-and-ads/pull/1868/files#r1094641305

**Comment by puntope** [comment]
Behaviour tested:

- I can add the new portrait image with the right ratio ✅ 
- When I send wrong ratio I get a descriptive error ✅ 
- I can get portrait image suggestions from a post with featured image 1200x1200 got ✅ 

```
 "portrait_marketing_image": [
        "https://wp-woo.test/wp-content/uploads/woocommerce-placeholder-960x1200.png"
 ],
```
⚠️ One thing I notice is about when the number of portrait images is 5. The adding 1 extra gives me success... However the image has not been updated. I guess maybe better to return error.

**Review by puntope — COMMENTED** [review]
Howdy @jorgemd24, thanks for this improvement. I reviewed the PR. 

Only one blocker from my side regarding the issue when maximum number of images is reached and attempting to update doesn't return error. 

Thanks

**Comment by jorgemd24** [comment]
Hi @puntope,

Thanks for reviewing this. I added your suggestions. 

> ⚠️ One thing I notice is about when the number of portrait images is 5. The adding 1 extra gives me success... However the image has not been updated. I guess maybe better to return error.

You can add more than 5 portrait images until 20, but the total of marketing images: square + landscape + portrait can not exceed 20. Keep in mind that if you add the same image, the Ads API will ignore it as it checks if the asset already exists. 

The limit of 5 is only for the suggestions, so we make a total of 20 marketing images. (10 landscape + 5 square + 5 portrait)

**Comment by puntope** [comment]
>  Keep in mind that if you add the same image, the Ads API will ignore it as it checks if the asset already exist

Thats why :) Thanks for the explanation. I was able to add the 6th image

**Review by puntope — APPROVED** [review]

**Pull request merged**